internal/expr/query.go

```diff
@@ -36,92 +36,120 @@ func (pe *PreparedExpr) Query(args ...any) (ce *QueryExpr, err error) {
-
-	for i, in := range pe.inputs {
-		v, ok := typeValue[in.structType]
+	for i, typeElement := range pe.inputs {
+		v, ok := typeValue[typeElement.outerType()]
 		if !ok {
-			return nil, fmt.Errorf(`type %s not found, have: %s`, in.structType.Name(), strings.Join(typeNames, ", "))
+			return nil, fmt.Errorf(`type %s not found, have: %s`, typeElement.outerType().Name(), strings.Join(typeNames, ", "))
+		}
+		switch te := typeElement.(type) {
+		case field:
+			qargs = append(qargs, sql.Named("sqlair_"+strconv.Itoa(i), v.FieldByIndex(te.index).Interface()))
+		case mapKey:
+			// MapIndex returns a zero value of the key is not in the map so we
+			// need to check the MapKeys().
+			var val reflect.Value
+			for _, key := range v.MapKeys() {
+				if key.String() == te.name {
+					val = v.MapIndex(reflect.ValueOf(te.name))
+					break
+				}
+			}
+			if !val.IsValid() {
+				return nil, fmt.Errorf(`map does not contain key %q`, te.name)
+			}
+			qargs = append(qargs, sql.Named("sqlair_"+strconv.Itoa(i), val.Interface()))
 		}
-		named := sql.Named("sqlair_"+strconv.Itoa(i), v.FieldByIndex(in.index).Interface())
-		qargs = append(qargs, named)
 	}
-
 	return &QueryExpr{outputs: pe.outputs, sql: pe.sql, args: qargs}, nil
 }
 
+type MapDecodeInfo struct {
+	m         reflect.Value
+	valuePtrs []any
+	keyIndex  map[string]int
+}
+
 // ScanArgs returns list of pointers to the struct fields that are listed in qe.outputs.
 // All the structs mentioned in the query must be in outputArgs.
 // All outputArgs must be structs.
-func (qe *QueryExpr) ScanArgs(columns []string, outputArgs []any) ([]any, error) {
+func (qe *QueryExpr) ScanArgs(columns []string, outputArgs []any) ([]any, []*MapDecodeInfo, error) {
+	var typesInQuery = []string{}
+	var inQuery = make(map[reflect.Type]bool)
+	for _, typeElement := range qe.outputs {
+		if ok := inQuery[typeElement.outerType()]; !ok {
+			inQuery[typeElement.outerType()] = true
+			typesInQuery = append(typesInQuery, typeElement.outerType().Name())
+		}
+	}
+
+	var mapDecodeInfos = make(map[reflect.Type]*MapDecodeInfo)
+	var typeDest = make(map[reflect.Type]reflect.Value)
 	outputVals := []reflect.Value{}
 	for _, outputArg := range outputArgs {
 		if outputArg == nil {
-			return nil, fmt.Errorf("need pointer to struct, got nil")
+			return nil, nil, fmt.Errorf("need map or pointer to struct, got nil")
 		}
 		outputVal := reflect.ValueOf(outputArg)
-		if outputVal.Kind() != reflect.Pointer {
-			return nil, fmt.Errorf("need pointer to struct, got %s", outputVal.Kind())
+		k := outputVal.Kind()
+		if k != reflect.Map {
+			if k != reflect.Pointer {
+				return nil, nil, fmt.Errorf("need map or pointer to struct, got %s", k)
+			}
+			if outputVal.IsNil() {
+				return nil, nil, fmt.Errorf("got nil pointer")
+			}
+			outputVal = outputVal.Elem()
+			k = outputVal.Kind()
+			if k != reflect.Struct && k != reflect.Map {
+				return nil, nil, fmt.Errorf("need map or pointer to struct, got pointer to %s", k)
```

Comment:
Should this instead say "need pointer to map or struct, got pointer to %s" ?

Reply:
Maps can be passed as a value, since they are themsevles a pointer under the hood, so they can pass one as a pointer if they want, but also by value.